mlir/include/mlir/Bindings/Python/PybindAdaptors.h

```diff
@@ -495,6 +495,8 @@ class mlir_type_subclass : public pure_subclass {
           .attr("replace")(superCls.attr("__name__"), captureTypeName);
     });
     if (getTypeIDFunction) {
+      def_staticmethod("get_static_typeid",
+                       [getTypeIDFunction]() { return getTypeIDFunction(); });
```

Comment:
This should probably better be `def_property_readonly_static(...` as defined in Pybind and used by `PyConcreteType` [here](https://github.com/llvm/llvm-project/blob/fd1ddd1bca4d43c49862274c12981b36e7a1ffe8/mlir/lib/Bindings/Python/IRModule.h#L950), but I am missing context to reimplement this.

Reply:
Ah now I remember why I didn't forward - because I couldn't get all the way to implementing `def_property_readonly_static` up above. Lemme try again.

Reply:
@martin-luecke can you try this patch

```diff
diff --git a/mlir/include/mlir/Bindings/Python/PybindAdaptors.h b/mlir/include/mlir/Bindings/Python/PybindAdaptors.h
index 125f9b92937b..e0fd338f7109 100644
--- a/mlir/include/mlir/Bindings/Python/PybindAdaptors.h
+++ b/mlir/include/mlir/Bindings/Python/PybindAdaptors.h
@@ -18,6 +18,7 @@
 #ifndef MLIR_BINDINGS_PYTHON_PYBINDADAPTORS_H
 #define MLIR_BINDINGS_PYTHON_PYBINDADAPTORS_H
 
+#include <pybind11/detail/internals.h>
 #include <pybind11/pybind11.h>
 #include <pybind11/pytypes.h>
 #include <pybind11/stl.h>
@@ -357,6 +358,17 @@ public:
     return *this;
   }
 
+  template <typename Func, typename... Extra>
+  pure_subclass &def_property_readonly_static(const char *name, Func &&f,
+                                              const Extra &...extra) {
+    py::cpp_function cf(std::forward<Func>(f), py::name(name), extra...);
+    auto *staticPropertyType = py::detail::get_internals().static_property_type;
+    auto builtinProperty =
+        py::reinterpret_borrow<py::object>((PyObject *)staticPropertyType);
+    thisClass.attr(name) = builtinProperty(cf);
+    return *this;
+  }
+
   template <typename Func, typename... Extra>
   pure_subclass &def_classmethod(const char *name, Func &&f,
                                  const Extra &...extra) {
@@ -495,8 +507,9 @@ public:
           .attr("replace")(superCls.attr("__name__"), captureTypeName);
     });
     if (getTypeIDFunction) {
-      def_staticmethod("get_static_typeid",
-                       [getTypeIDFunction]() { return getTypeIDFunction(); });
+      def_property_readonly_static("get_static_typeid", [getTypeIDFunction]() {
+        return getTypeIDFunction();
+      });
       py::module::import(MAKE_MLIR_PYTHON_QUALNAME("ir"))
           .attr(MLIR_PYTHON_CAPI_TYPE_CASTER_REGISTER_ATTR)(
               getTypeIDFunction())(pybind11::cpp_function(
```

Reply:
This approach has quite a rats tail attached to it because it seems to change the behavior of how other fields are handled. This leads to lots of errors like `AttributeError: 'pybind11_static_property' object attribute '__doc__' is read-only`. 
Pybind uses [this code here](https://github.com/pybind/pybind11/blob/7d538a42750c8580eeaac10e505840a3694b04c8/include/pybind11/detail/class.h#L539) to work around this behavior. 
I will for now leave this as a static function to not introduce more complexity to `pure_subclass` just for this.